Mathlib/Computability/GNFA.lean

```diff
@@ -0,0 +1,276 @@
+/-
+Copyright (c) 2022 Russell Emerine, 2024 Tom Kranz. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Russell Emerine, Tom Kranz
+-/
+import Mathlib.Computability.RegularExpressions
+import Mathlib.Data.FinEnum.Option
+
+/-!
+# Generalized Nondeterministic Finite Automata
+
+This file contains the definition of a Generalized Nondeterministic Finite Automaton, a state
+machine which determines whether a string (implemented as a `List` over an arbitrary alphabet) is in
+a regular set by evaluating the string over its net of regular expressions. We show
+that GNFA's are equivalent to `NFA`'s, and that GNFA's are equivalent to smaller GNFA's with a state
+"ripped" out. Through this mechanism, we show that `NFA`'s are equivalent to `RegularExpression`s.
+
+## References
+E.g. [Michael Sipser's *Introduction to the Theory of Computation*, Lemma 1.32][zbMATH05595151]
+-/
+
+
+universe u v
+
+/-- A GNFA is a set of `|σ| + 2` states and a transition function between two states. The transition
+function takes the starting state (represented by `Option.none` in the first argument) or any
+internal state (consistently represented by `Option.some`s) as the first state, and the accepting
+state (represented by `Option.none` in the second argument) or any internal state as the second
+state. There is a transition between *all* of these
+combinations, in the form of a `RegularExpression`. When following a transition, some matching
+prefix of the input string is taken. What would be a missing transition in ordinary automata can be
+achieved by yielding a `RegularExpression` matching `∅`; canonically
+`RegularExpression.zero`.
+-/
+-- Porting note: removed Fintype instance for σ as an argument
+structure GNFA (α : Type u) (σ : Type v) where
+  /-- Yields the regular expression (sub)strings have to match to transition from the first to the
+  second argument state. `Option.none` in the first position signifies a transition originating in
+  the start state, `Option.none` in the second position signifies a transition targetting the
+  accept state. -/
+  step : Option σ → Option σ → RegularExpression α
+
+variable {α : Type u} {σ : Type v}
+
+namespace GNFA
+
+/-- The GNFA admitting only impassable transitions (and be it only from start to accept state)
+will always be available. -/
+instance : Inhabited (GNFA α σ) :=
+  ⟨GNFA.mk fun _ _ ↦ 0⟩
+
+/-- A `trace` of an internal state and a string of a GNFA represents the possibility of getting to
+the state via transitions that match the elements of some partitioning of the string.
+This can be interpreted as the `Set` of strings that reach a certain inner state;
+i.e. its `Language`.
+-/
+inductive trace (M : GNFA α σ) : σ → Language α
+  | start : ∀ {x} q, x ∈ (M.step none (some q)).matches' → M.trace q x
+  | step : ∀ {x y z} p q,
+    y ∈ M.trace p → z ∈ (M.step (some p) (some q)).matches' → x = y ++ z → M.trace q x
```

Comment:
Please be consistent with whether you write `M.trace p x` or `x ∈ M.trace p`. If you decide to use the former, you should also write `(M.step (some p) (some q)).matches' z` instead of `z ∈ (M.step (some p) (some q)).matches'`.